File: 8_kyu/Heads_and_Legs.lua
local kata = {}

function kata.animals(heads, legs)
  -- It is:
  -- heads = cows + chickens
  -- legs = 4*cows + 2*chickens
  -- solve this system of 2 linear equations
  local cows = legs/2 - heads
  local chickens = 2*heads - legs/2
  if cows < 0 or chickens < 0 or legs%2 ~= 0 then
    return 'No solutions'
  else
    return { chickens, cows }
  end
end

return kata

File: 8_kyu/Heads_and_Legs.py
def animals(heads: int, legs: int) -> tuple[int,int]|str:
    """
    Calculates number of chickens and cows on the farm.
    heads -- Count of all heads
    legs -- Count of all legs
    return -- Tuple of #chickens, #cows, or error message in string forma.
    """
    if not (legs >= 0 and legs%2 == 0 and legs//4 <= heads <= legs//2):
        return "No solutions"
    cows = legs//2 - heads
    chickens = heads - cows
    return chickens, cows
